extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxConfigBuilder.java

```diff
@@ -14,14 +16,21 @@ public SmallRyeConfigBuilder configBuilder(final SmallRyeConfigBuilder builder)
         if (builder.getDefaultValues().get(QUARKUS_HTTP_HOST) == null) {
             // Sets the default host config value, depending on the launch mode
             if (LaunchMode.isRemoteDev()) {
-                // in remote-dev mode we need to listen on all interfaces
+                // in remote dev mode, we want to listen on all interfaces
+                // to make sure the application is accessible
                 builder.withDefaultValue(QUARKUS_HTTP_HOST, ALL_INTERFACES);
+            } else if (LaunchMode.current().isDevOrTest()) {
+                if (!isWSL()) {
+                    // in dev mode, we want to listen only on localhost
+                    // to make sure the app is not accessible from the outside
+                    builder.withDefaultValue(QUARKUS_HTTP_HOST, LOCALHOST);
+                } else {
+                    // except when using WSL, as otherwise the app wouldn't be accessible from the host
+                    builder.withDefaultValue(QUARKUS_HTTP_HOST, ALL_INTERFACES);
+                }
             } else {
-                // In dev-mode we want to only listen on localhost so others on the network cannot connect to the application.
-                // However, in WSL this would result in the application not being accessible,
-                // so in that case, we launch it on all interfaces.
-                builder.withDefaultValue(QUARKUS_HTTP_HOST,
-                        (LaunchMode.current().isDevOrTest() && !isWSL()) ? "localhost" : ALL_INTERFACES);
+                // in all the other cases, we make sure the app is accessible on all the interfaces by default
+                builder.withDefaultValue(QUARKUS_HTTP_HOST, ALL_INTERFACES);
             }
         }
         return builder;
```

Comment:
@dmlloyd could you have a look at this change as it could have some security consequences.

I adjusted it because the previous commit by the OP actually changed the default value and I wanted to make it extra clear (even if a bit more verbose).